File: 2015/deno/day05/part-2.ts
export function processPart2(input: string): number {
	const trimmed = input.trim();
	const lines = trimmed.split('\n');

	let niceStrings = 0;

	for (const line of lines) {
		if (isNice(line)) {
			niceStrings++;
		}
	}
	return niceStrings;
}

function isNice(s: string): boolean {
	if (!twoPairs(s)) {
		return false;
	}
	if (!sepRepeat(s)) {
		return false;
	}
	return true;
}

function twoPairs(s: string): boolean {
	const pairs = new Map<string, number>();
	for (let i = 0; i < s.length - 1; i++) {
		const pair = s.substring(i, i + 2);
		if (pairs.has(pair)) {
			if (i - pairs.get(pair)! > 1) {
				return true;
			}
		} else {
			pairs.set(pair, i);
		}
	}
	return false;
}

function sepRepeat(s: string): boolean {
	for (let i = 2; i < s.length; i++) {
		if (s[i] === s[i - 2]) {
			return true;
		}
	}
	return false;
}

File: 2015/go/day05/part2.go
package main

import "strings"

func processPart2(input string) int {
	lines := strings.Split(input, "\n")
	niceStrings := 0

	for _, line := range lines {
		if isNiceString(line) {
			niceStrings++
		}
	}

	return niceStrings
}

func isNiceString(s string) bool {
	if !splitPair(s) {
		return false
	}
	if !twoPairs(s) {
		return false
	}
	return true
}

func splitPair(s string) bool {
	for i := 2; i < len(s); i++ {
		if s[i] == s[i-2] {
			return true
		}
	}
	return false
}

func twoPairs(s string) bool {
	pairs := make(map[string][2]int)

	for i, j := 0, 1; j < len(s); i, j = i+1, j+1 {
		pair := s[i : j+1]
		if indices, exists := pairs[pair]; exists {
			if indices[1] < i {
				return true
			}
		} else {
			pairs[pair] = [2]int{i, j}
		}

	}

	return false
}
